tsdb/engine/tsm1/cache.go

```diff
@@ -0,0 +1,320 @@
+// it will evict older data as necessary to make room for the new entries.
+type Cache struct {
+	mu         sync.RWMutex
+	store      map[string]entries
+	checkpoint uint64
+	size       uint64
+	maxSize    uint64
+
+	lru *lru // List of entry keys from most recently accessed to least.
+}
+
+// NewCache returns an instance of a cache which will use a maximum of maxSize bytes of memory.
+func NewCache(maxSize uint64) *Cache {
+	return &Cache{
+		maxSize: maxSize,
+		store:   make(map[string]entries),
+		lru:     newLRU(),
+	}
+}
+
+// WriteKey writes the set of values for the key to the cache. It associates the data with
+// the given checkpoint. This function is goroutine-safe.
+//
+// TODO: This function is a significant potential bottleneck. It is possible that keys could
+// be modified while an eviction process was taking place, so a big cache-level lock is in place.
+// Need to revisit this. It's correct but may not be performant (it is the same as the existing
+// design however)
+func (c *Cache) Write(key string, values []Value, checkpoint uint64) error {
+	c.mu.Lock()
+	defer c.mu.Unlock()
+	if checkpoint < c.checkpoint {
+		return ErrCacheInvalidCheckpoint
+	}
+
+	newSize := c.size + uint64(Values(values).Size())
+	if newSize >= c.maxSize {
+		c.evict(newSize - c.maxSize)
+	}
+
+	// Size OK now?
+	if c.size >= c.maxSize {
+		return ErrCacheMemoryExceeded
+	}
+
+	e, ok := c.store[key]
+	if !ok {
+		e = newEntries()
+		c.store[key] = e
+	}
+	e.add(values, checkpoint)
+	c.size = newSize
+
+	// Mark entry as most-recently used.
+	c.lru.MoveToFront(key)
+
+	return nil
+}
+
+// SetCheckpoint informs the cache that updates received up to and including checkpoint
+// can be safely evicted. Setting a checkpoint does not mean that eviction up to that
+// point will actually occur.
+func (c *Cache) SetCheckpoint(checkpoint uint64) error {
+	c.mu.Lock()
+	defer c.mu.Unlock()
+	if checkpoint < c.checkpoint {
+		return ErrCacheInvalidCheckpoint
+	}
+	c.checkpoint = checkpoint
+	return nil
+}
+
+// Size returns the number of bytes the cache currently uses.
+func (c *Cache) Size() uint64 {
+	c.mu.RLock()
+	defer c.mu.RUnlock()
+	return c.size
+}
+
+// MaxSize returns the maximum number of bytes the cache may consume.
+func (c *Cache) MaxSize() uint64 {
+	return c.maxSize
+}
+
+// Keys returns a sorted slice of all keys under management by the cache.
+func (c *Cache) Keys() []string {
+	var a []string
+	for k, _ := range c.store {
+		a = append(a, k)
+	}
+	sort.Strings(a)
+	return a
+}
+
+// Checkpoint returns the current checkpoint for the cache.
+func (c *Cache) Checkpoint() uint64 {
+	c.mu.RLock()
+	defer c.mu.RUnlock()
+	return c.checkpoint
+}
+
+// Evict instructs the cache to evict.
+func (c *Cache) Evict(size uint64) uint64 {
+	c.mu.Lock()
+	defer c.mu.Unlock()
+	return c.evict(size)
+}
+
+// Cursor returns a cursor for the given key.
+func (c *Cache) Cursor(key string) tsdb.Cursor {
```

Comment:
Instead of returning a `Cursor` here, we should just return a `Values` slice that is clone of the current `Values` for the key (and possibly timerange).  We have way too many cursor implementations and we can design this so we only have one for the TSM engine.